### PR TITLE
Show CFGs on Nightly

### PR DESCRIPTION
**Pull request by ajpal**
image

image
image

### CODE REVIEW

File: infra/nightly.sh [hunk 1]
```diff
@@ -62,6 +62,9 @@ fi
 # Generate latex after running the profiler (depends on profile.json)
 ./infra/generate_line_counts.py "$NIGHTLY_DIR"
 
+# Generate CFGs for LLVM after running the profiler
+./infra/generate_cfgs.py "$NIGHTLY_DIR/data/llvm"
```
Comment (oflatt) [inline infra/nightly.sh:66]: Might be better for the `profile.py` file to call `generate_cfgs` directly- less bash communication is better

File: infra/profile.py [hunk 2]
```diff
@@ -103,8 +103,9 @@ def should_have_llvm_ir(runMethod):
 def dump_llvm_ir(runMethod, benchmark, output_directory):
```
Comment (oflatt) [inline infra/profile.py:103]: What is the point of `dump_llvm_ir`? It seems to move .ll files from one folder to another. Shouldn't we just put it in that folder in the first place?

**Review by oflatt — CHANGES_REQUESTED** [review]
This code requires lots of mutating the file system, so I'm getting confused about all the folders. If we can minimize the number of places we specify paths and minimize moving files around that would be great.

Besides that, making clear wrappers for mutating the state of the html page would be good

File: infra/nightly.sh (hunk 1 above)
Comment (ajpal) [reply infra/nightly.sh:66]: I kind of disagree? I think the separation is good- `profile.py` is in charge of running all the benchmarks and stuff, `generate_cfgs` is in charge of reading that data in and writing pngs out.

File: infra/nightly-resources/handlers.js [hunk 3]
```diff
@@ -99,39 +100,45 @@ async function loadBaseline(url) {
...
-  GLOBAL_DATA.chart.mode = elt.value;
-  document.getElementById("speedup-formula").style.visibility =
-    elt.value === "speedup" ? "visible" : "hidden";
-  refreshChart();
-}
-
-function copyToClipboard(eltId) {
-  navigator.clipboard.writeText(document.getElementById(eltId).innerText);
+function toggleAllPngs(elt) {
+  const btns = Array.from(document.getElementsByTagName("button"));
```
[the remark this answers is not in the document]
Comment (ajpal) [reply infra/nightly-resources/handlers.js]: Changed to getting elements by class name

File: infra/nightly-resources/llvm.js [hunk 4]
```diff
@@ -0,0 +1,43 @@
+async function showIR(benchmark, runMode) {
+  const llvm = await fetchText(`./data/llvm/${benchmark}/${runMode}/llvm.ll`);
+  document.getElementById("llvm-ir").innerText = llvm;
+}
+
+async function showCFGs(benchmark, runMode) {
+  let pngs = (
+    await fetchText(`./data/llvm/${benchmark}/${runMode}/png_names.txt`)
```
[the remark this answers is not in the document]
Comment (ajpal) [reply infra/nightly-resources/llvm.js:10]: Discussed offline- this is the least bad option :)

File: infra/profile.py (hunk 2 above)
Comment (ajpal) [reply infra/profile.py]: Fixed

File: infra/generate_cfgs.py [hunk 5]
```diff
@@ -0,0 +1,59 @@
+#!/usr/bin/env python3
+import glob
+import os
+
+def make_cfgs(bench, data_dir):
+  cwd = os.getcwd()
+  path = f"{data_dir}/{bench}"
+  runmodes = os.listdir(path)
+  for mode in runmodes:
+    os.chdir(f"{path}/{mode}")
```
Comment (oflatt) [inline infra/generate_cfgs.py:10]: Sorry I didn't catch this, but I think it's bad practice to change the working directory
Can we build the relative path everywhere instead?

File: infra/generate_cfgs.py [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+    else:
+      opt = "opt"
+
+    # https://llvm.org/docs/Passes.html#dot-cfg-print-cfg-of-function-to-dot-file
+    cmd = f"{opt} -disable-output -passes=dot-cfg {bench}-{mode}.ll"
+    os.system(cmd)
+
+    # Delete the -init.ll file (We don't need it for nightly,
+    # so just reduce the amount of clutter we copy to the nightly machine)
+    os.system(f"rm {bench}-{mode}-init.ll")
```
Comment (oflatt) [inline infra/generate_cfgs.py:27]: Should this be deleted when it is created instead of here?

File: infra/profile.py [hunk 7]
```diff
@@ -129,17 +123,19 @@ def aggregate(compile_times, bench_times, output_directory):
     print("Usage: profile.py <bril_directory> <output_directory>")
     exit(1)
 
-  profile_path, output_path = os.sys.argv[1:]
+  # Create tmp directory for intermediate files
+  os.mkdir(TMP_DIR)
```
Comment (oflatt) [inline infra/profile.py:127]: If this script is killed and the tmp dir is still there, does this fail? Do we want to delete it in that case?

**Review by oflatt — APPROVED** [review]
Looks great! Minor comments, not blocking

File: infra/generate_cfgs.py (hunk 5 above)
Comment (ajpal) [reply infra/generate_cfgs.py:10]: Then all the dot and png files would get created in the top-level dir, which is not what we want

**Pull request merged**